scipy/stats/_fit.py

```diff
@@ -565,3 +566,528 @@ def nllf(free_params, data=data):  # bind data NOW
+        pvalue : float
+            The proportion of elements in the null distribution with
+            statistic values at least as extreme as the statistic value of the
+            provided `data`.
+        null_distribution : ndarray
+            The value of the statistic for each Monte Carlo sample
+            drawn from the null-hypothesized distribution.
+
+    Notes
+    -----
+    This is a generalized Monte Carlo goodness-of-fit procedure, special cases
+    of which correspond with various Anderson-Darling tests, Lilliefors' test,
+    etc.
+
+    Traditionally, critical values corresponding with a set a fixed set of
+    significance levels are pre-calculated using Monte Carlo methods. Users
+    perform the test by calculating the value of the test statistic only for
+    their observed `data` and comparing this value to tabulated critical
+    values. This practice is not very flexible, as tables are not available for
+    all distributions and combinations of known and unknown parameter values.
+    Also, results can be inaccurate when critical values are interpolated from
+    limited tabulated data to correspond with the user's sample size and
+    fitted parameter values. To overcome these shortcomings, this function
+    allows the user to perform the Monte Carlo trials adapted to their
+    particular data.
+
+    First, any unknown parameters of the distribution family specified by
+    `dist` are fit to the provided `data`. These values of the parameters
+    specify a particular member of the distribution family referred to as the
+    "null-hypothesized distribution", that is, the distribution from which the
+    data were sampled under the null hypothesis. The `statistic`, which
+    compares data to a distribution, is computed between `data` and the
+    null-hypothesized distribution.
+
+    Next, many (specifically `n_mc_samples`) new samples, each containing the
+    same number of observations as `data`, are drawn from the
+    null-hypothesized distribution. Any unknown parameters of the distribution
+    family `dist` are fit to *each resample*, and the `statistic` is computed
+    between each sample and its corresponding fitted distribution. These
+    values of the statistic form the Monte Carlo null distribution (not to be
+    confused with the "null-hypothesized distribution" above).
+
+    The p-value of the test is the proportion of statistic values in the Monte
+    Carlo null distribution that are at least as extreme as the statistic value
+    of the provided `data`. More precisely, the p-value is given by
+
+    .. math::
+
+        p = \frac{b + 1}
+                 {m + 1}
+
+    where :math:`b` is the number of statistic values in the Monte Carlo null
+    distribution that are greater than or equal to the the statistic value
+    calculated for `data`, and :math:`m=` is the number of elements in the
+    Monte Carlo null distribution (`n_mc_samples`). The addition of :math:`1`
```

Comment:
reminder for myself: check reference that explains the addition of 1

Reply:
The reference is very interesting. When I was writing `monte_carlo_test` and `permutation_test` (without exhaustive resampling of all possible "permutations"), I thought of the p-values as approximations of exact p-values. Instead, you can get "exact" p-values even with randomized tests in the sense that they can exactly control the long-run Type I error rate. It just requires a little care. For Monte Carlo tests, it turns out that all it takes to get such an exact p-value is to add one to the numerator and denominator. For randomized permutation tests in which resamples are drawn with replacement (i.e. i could resample the same "permutation" again), adding +1 to the top and bottom is an imperfect (but quite good) bias correction. 

Someday I'll add the calculation for the exact p-value from these tests or - even better - change the code to draw resamples without replacement (so that adding +1 to the numerator and denominator results in an exact test). I think I can do this for the paired-sample and correlation test permutations, but the independent sample permutations would be harder.